stats/opentelemetry/internal/grpc_trace_bin_propagator.go

```diff
@@ -0,0 +1,140 @@
+
+// GRPCTraceBinPropagator is an OpenTelemetry TextMapPropagator which is used
+// to extract and inject trace context data from and into messages exchanged by
+// gRPC applications. It propagates trace data in binary format using the
+// 'grpc-trace-bin' header.
+type GRPCTraceBinPropagator struct{}
+
+// Inject sets OpenTelemetry trace context information from the Context into
+// the carrier.
+//
+// If the carrier is a CustomCarrier, trace data is directly injected in a
+// binary format using the 'grpc-trace-bin' header (fast path). Otherwise,
+// the trace data is base64 encoded and injected using the same header in
+// text format (slow path).
+func (p GRPCTraceBinPropagator) Inject(ctx context.Context, carrier otelpropagation.TextMapCarrier) {
+	span := oteltrace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return
+	}
+
+	bd := Binary(span.SpanContext())
+	if bd == nil {
+		return
+	}
+
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		cc.SetBinary(bd)
+		return
+	}
+	carrier.Set(itracing.GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(bd))
+}
+
+// Extract reads OpenTelemetry trace context information from the carrier into a
+// Context.
+//
+// If the carrier is a CustomCarrier, trace data is read directly in a binary
+// format from the 'grpc-trace-bin' header (fast path). Otherwise, the trace
+// data is base64 decoded from the same header in text format (slow path).
+func (p GRPCTraceBinPropagator) Extract(ctx context.Context, carrier otelpropagation.TextMapCarrier) context.Context {
+	var bd []byte
+
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		bd = cc.GetBinary()
+	} else {
+		bd, _ = base64.StdEncoding.DecodeString(carrier.Get(itracing.GRPCTraceBinHeaderKey))
+	}
+	if bd == nil {
+		return ctx
+	}
+
+	sc, ok := FromBinary([]byte(bd))
+	if !ok {
+		return ctx
+	}
+
+	return oteltrace.ContextWithRemoteSpanContext(ctx, sc)
+}
+
+// Fields returns the keys whose values are set with Inject.
+//
+// GRPCTraceBinPropagator always returns a slice containing only
+// `grpc-trace-bin` key because it only sets the 'grpc-trace-bin' header for
+// propagating trace context.
+func (p GRPCTraceBinPropagator) Fields() []string {
+	return []string{itracing.GRPCTraceBinHeaderKey}
+}
+
+// Binary returns the binary format representation of a SpanContext.
+//
+// If sc is the zero value, returns nil.
+func Binary(sc oteltrace.SpanContext) []byte {
+	if sc.Equal(oteltrace.SpanContext{}) {
+		return nil
+	}
+	var b [29]byte
+	traceID := oteltrace.TraceID(sc.TraceID())
+	copy(b[2:18], traceID[:])
+	b[18] = 1
+	spanID := oteltrace.SpanID(sc.SpanID())
+	copy(b[19:27], spanID[:])
+	b[27] = 2
+	b[28] = uint8(oteltrace.TraceFlags(sc.TraceFlags()))
+	return b[:]
+}
+
+// FromBinary returns the SpanContext represented by b.
+//
+// If b has an unsupported version ID or contains no TraceID, FromBinary
+// returns with zero value SpanContext and false.
+func FromBinary(b []byte) (oteltrace.SpanContext, bool) {
```

Comment:
Actually, as per [gRFC](https://github.com/grpc/proposal/blob/master/A72-open-telemetry-tracing.md#migration-to-opentelemetry-cross-process-networking-concerns)

`a grpc-trace-bin formatter implementation for OpenTelemetry is needed in each language, which can be similar to the OpenCensus implementation.` 

So, this implementation is directly taken from OpenCensus Go [FromBinary](https://github.com/census-instrumentation/opencensus-go/blob/01e6da5fc01c42aca1e0ce315f41744876e9fcbb/trace/propagation/propagation.go#L74). The only difference is semantics of otel while copying values.

gRFC also calls out that this is a temporary solution until open telemetry provides a similar formatter which we can use directly.

Reply:
Java has also done the same https://github.com/grpc/grpc-java/pull/11409/files. Here is the opencensus java implementation https://github.com/census-instrumentation/opencensus-java/blob/baa68680b19040ff7739e314a2422a25cf41e27a/impl_core/src/main/java/io/opencensus/implcore/trace/propagation/BinaryFormatImpl.java#L98

Reply:
Just because another codebase does it a certain way doesn't mean it's necessarily correct or the best style. Please fix this. When we moved some OpenCensus code to our repository we rewrote a lot of operations.

Reply:
I have rewritten `FromBinary` similar to c++ way. I have also made them unexported as I think we don't want these to be used directly. I have also added separate tests for both FromBinary and Binary since we have some edge cases.